docs/source/examples/vqe-pyquil-demo.myst

```diff
@@ -0,0 +1,312 @@
+
+```{code-cell} ipython3
+hamiltonian = sZ(0)
+pauli_sum = PauliSum([hamiltonian])
+
+for j, term in enumerate(pauli_sum.terms):
+    meas_basis_change = Program()
+    marked_qubits = []
+    for index, gate in term:
+        marked_qubits.append(index)
+        if gate == "X":
+            meas_basis_change.inst(RY(-np.pi / 2, index))
+        elif gate == "Y":
+            meas_basis_change.inst(RX(np.pi / 2, index))
+        program += meas_basis_change
+    readout_qubit = program.declare("ro", "BIT", max(marked_qubits) + 1)
+
+samples = 3000
+program.wrap_in_numshots_loop(samples)
+```
+
+Compute expectation value of the Hamiltonian over the over the distribution 
+generated from the quantum program. The following function is a modified version 
+of `expectation` from the `VQE` function in Grove [[1]](#References). Here the noisy 
+gates are defined inside the executor function via `add_noise_to_circuit`, since 
+pyQuil custom gates cannot be folded in Mitiq.
+
+```{code-cell} ipython3
+def executor(
+    thetas,
+    backend,
+    readout_qubit,
+    samples: int,
+    pauli_sum: Union[PauliSum, PauliTerm, np.ndarray],
+    pyquil_prog: Program,
+) -> float:
+    """
+    Compute the expectation value of pauli_sum over the distribution generated from 
+    pyquil_prog.
+    """
+    noisy = pyquil_prog.copy()
+    noisy += [
+        MEASURE(qubit, r) for qubit, r in zip(list(range(max(marked_qubits) + 1)), readout_qubit)
+    ]
+    noisy = add_noise_to_circuit(noisy)
+    expectation = 0.0
+    pauli_sum = PauliSum([pauli_sum])
+    for j, term in enumerate(pauli_sum.terms):
+        qubits_to_measure = []
+        for index, gate in term:
+            qubits_to_measure.append(index)
+            meas_outcome = expectation_from_sampling(
+                thetas, noisy, qubits_to_measure, backend, samples
+            )
+            expectation += term.coefficient * meas_outcome
+    return expectation.real
+```
+
+The following function is a modified version of `expectation_from_sampling` 
+from the `VQE` function in Grove [[1]](#References). It is modified to follow pyQuil 
+conventions for defining custom gates.
+
+```{code-cell} ipython3
+def expectation_from_sampling(
+    thetas, executable: Program, marked_qubits: List[int], backend, samples: int
+) -> float:
+    """Calculate the expectation value of the Zi operator where i ranges over all 
+    qubits given in marked_qubits.
+    """
+    bitstring_samples = backend.run(executable, memory_map={"theta": thetas})
+    bitstring_tuples = list(map(tuple, bitstring_samples))
+
+    freq = Counter(bitstring_tuples)
+
+    
+    exp_val = 0
+    for bitstring, count in freq.items():
+        bitstring_int = int("".join([str(x) for x in bitstring[::-1]]), 2)
+        if parity_even_p(bitstring_int, marked_qubits):
+            exp_val += float(count) / samples
+        else:
+            exp_val -= float(count) / samples
+    return exp_val
```

Comment:
Another FYI, no change requested - this logic is in `mitiq.Observable` now. https://github.com/unitaryfund/mitiq/blob/5ee3da4ba39a732060f700ae269f9ab0b2ca96a9/mitiq/observable/observable.py#L103-L105